tests/splines/extrapolation_rule.cpp

```diff
@@ -0,0 +1,403 @@
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <iosfwd>
+#include <vector>
+
+#include <experimental/mdspan>
+
+#include <ddc/ddc.hpp>
+#include <ddc/kernels/splines.hpp>
+
+#include <gtest/gtest.h>
+
+#include "ddc/coordinate.hpp"
+#include "ddc/detail/macros.hpp"
+#include "ddc/discrete_domain.hpp"
+#include "ddc/for_each.hpp"
+#include "ddc/uniform_point_sampling.hpp"
+
+#include "cosine_evaluator.hpp"
+#include "evaluator_2d.hpp"
+#include "polynomial_evaluator.hpp"
+#include "spline_error_bounds.hpp"
+
+#if defined(BC_PERIODIC)
+struct DimX
+{
+    static constexpr bool PERIODIC = false;
+};
+
+struct DimY
+{
+    static constexpr bool PERIODIC = true;
+};
+#elif defined(BC_GREVILLE)
+
+struct DimX
+{
+    static constexpr bool PERIODIC = false;
+};
+
+struct DimY
+{
+    static constexpr bool PERIODIC = false;
+};
+#endif
+
+static constexpr std::size_t s_degree = DEGREE;
+
+#if defined(BC_PERIODIC)
+static constexpr ddc::BoundCond s_bcl1 = ddc::BoundCond::GREVILLE;
+static constexpr ddc::BoundCond s_bcr1 = ddc::BoundCond::GREVILLE;
+static constexpr ddc::BoundCond s_bcl2 = ddc::BoundCond::PERIODIC;
+static constexpr ddc::BoundCond s_bcr2 = ddc::BoundCond::PERIODIC;
+#elif defined(BC_GREVILLE)
+static constexpr ddc::BoundCond s_bcl1 = ddc::BoundCond::GREVILLE;
+static constexpr ddc::BoundCond s_bcr1 = ddc::BoundCond::GREVILLE;
+static constexpr ddc::BoundCond s_bcl2 = ddc::BoundCond::GREVILLE;
+static constexpr ddc::BoundCond s_bcr2 = ddc::BoundCond::GREVILLE;
+#endif
+
+template <typename BSpX>
+using GrevillePoints = ddc::
+        GrevilleInterpolationPoints<BSpX, ddc::BoundCond::GREVILLE, ddc::BoundCond::GREVILLE>;
+
+#if defined(BSPLINES_TYPE_UNIFORM)
+template <typename X>
+using BSplines = ddc::UniformBSplines<X, s_degree>;
+
+// Gives discrete dimension. In the dimension of interest, it is deduced from the BSplines type. In the other dimensions, it has to be newly defined. In practice both types coincide in the test, but it may not be the case.
+template <typename X, typename I1, typename I2>
+using IDim = std::conditional_t<
+        std::disjunction_v<std::is_same<X, I1>, std::is_same<X, I2>>,
+        typename GrevillePoints<BSplines<X>>::interpolation_mesh_type,
+        ddc::UniformPointSampling<X>>;
+
+#elif defined(BSPLINES_TYPE_NON_UNIFORM)
+template <typename X>
+using BSplines = ddc::NonUniformBSplines<X, s_degree>;
+
+template <typename X, typename I1, typename I2>
+using IDim = std::conditional_t<
+        std::disjunction_v<std::is_same<X, I1>, std::is_same<X, I2>>,
+        typename GrevillePoints<BSplines<X>>::interpolation_mesh_type,
+        ddc::NonUniformPointSampling<X>>;
+#endif
```

Comment:
I would favor this implementation
```suggestion
#if defined(BSPLINES_TYPE_UNIFORM)
template <typename X>
using BSplines = ddc::UniformBSplines<X, s_degree>;

// Gives discrete dimension. In the dimension of interest, it is deduced from the BSplines type. In the other dimensions, it has to be newly defined. In practice both types coincide in the test, but it may not be the case.
template <typename X, typename I1, typename I2>
using IDim = std::conditional_t<
        (std::is_same_v<X, I1> || std::is_same_v<X, I2>),
        typename GrevillePoints<BSplines<X>>::interpolation_mesh_type,
        ddc::UniformPointSampling<X>>;

#elif defined(BSPLINES_TYPE_NON_UNIFORM)
template <typename X>
using BSplines = ddc::NonUniformBSplines<X, s_degree>;

template <typename X, typename I1, typename I2>
using IDim = std::conditional_t<
        (std::is_same_v<X, I1> || std::is_same_v<X, I2>),
        typename GrevillePoints<BSplines<X>>::interpolation_mesh_type,
        ddc::NonUniformPointSampling<X>>;
#endif
```